calabash-cucumber/lib/calabash-cucumber/core.rb

```diff
@@ -462,6 +465,192 @@ def pinch(in_out, options={})
         launcher.gesture_performer.pinch(in_out.to_sym,options)
       end
 
+      # Use keyboard to enter a character.
+      #
+      # @note
+      #  There are several special 'characters', some of which do not appear on
+      #  all keyboards; e.g. `Delete`, `Return`.
+      #
+      # @see #keyboard_enter_text
+      #
+      # @note
+      #  You should prefer to call `keyboard_enter_text`.
+      #
+      # @raise [RuntimeError] If there is no visible keyboard
+      # @raise [RuntimeError] If the keyboard (layout) is not supported
+      #
+      # @param [String] char The character to type
+      # @param [Hash] options Controls the behavior of the method.
+      # @option opts [Numeric] :wait_after_char (0.05) How long to sleep after
+      #  typing a character.
+      def keyboard_enter_char(char, options={})
+        expect_keyboard_visible!
+
+        default_opts = {:wait_after_char => 0.05}
+        merged_options = default_opts.merge(options)
+
+        special_char = launcher.gesture_performer.char_for_keyboard_action(char)
+
+        if special_char
+          launcher.gesture_performer.enter_char_with_keyboard(special_char)
+        elsif char.length == 1
+          launcher.gesture_performer.enter_char_with_keyboard(char)
+        else
+          raise ArgumentError, %Q[
+Expected '#{char}' to be a single character or a special string like:
+
+* Return
+* Delete
+
+To type strings with more than one character, use keyboard_enter_text.
+]
+        end
+
+        duration = merged_options[:wait_after_char]
+        if duration > 0
+          Kernel.sleep(duration)
+        end
+
+        []
+      end
+
+      # Touches the keyboard action key.
+      #
+      # The action key depends on the keyboard.  Some examples include:
+      #
+      # * Return
+      # * Next
+      # * Go
+      # * Join
+      # * Search
+      #
+      # @note
+      #  Not all keyboards have an action key.  For example, numeric keyboards
+      #  do not have an action key.
+      #
+      # @raise [RuntimeError] If the keyboard is not visible.
+      def tap_keyboard_action_key
+        expect_keyboard_visible!
+        launcher.gesture_performer.tap_keyboard_action_key
+      end
+
+      # Touches the keyboard delete key.
+      #
+      # @raise [RuntimeError] If the keyboard is not visible.
+      def tap_keyboard_delete_key
+        expect_keyboard_visible!
+        launcher.gesture_performer.tap_keyboard_delete_key
+      end
+
+      # Uses the keyboard to enter text.
+      #
+      # @param [String] text the text to type.
+      # @raise [RuntimeError] If the keyboard is not visible.
+      def keyboard_enter_text(text)
+        expect_keyboard_visible!
+        existing_text = text_from_first_responder
+        if existing_text && existing_text == ""
```

Comment:
Do we need this if/else? Seems like `text_from_first_responder` cannot return nil either way and the gsub should leave the string unaltered if existing_text = ""